src/Symfony/Component/DependencyInjection/Loader/XmlFileLoader.php

```diff
@@ -199,11 +202,47 @@ private function parseDefinition(\DOMElement $service, $file, array $defaults =
             } elseif (isset($defaults['public'])) {
                 $public = $defaults['public'];
             }
-            $this->container->setAlias((string) $service->getAttribute('id'), new Alias($alias, $public));
+            $this->container->setAlias($id, new Alias($alias, $public));
 
             return;
         }
 
+        $definition = $this->getDefinition($service, $file, $defaults);
+        if (null === $instanceof) {
+            return $definition;
+        }
+
+        $className = $definition->getClass() ?: $id;
```

Comment:
IIUC, it cannot be fixed in the loader directly because all parameters may have not been set at this time...